extensions/resource-deployment/src/helpers/optionSources.ts

```diff
@@ -0,0 +1,108 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as arc from 'arc';
+import { CategoryValue } from 'azdata';
+import { IOptionsSource } from '../interfaces';
+import * as loc from '../localizedConstants';
+import { apiService } from '../services/apiService';
+import { throwUnless } from '../utils';
+import { CacheManager } from './cacheManager';
+
+
+export type OptionsSourceType = 'ArcControllersOptionsSource';
+
+const OptionsSources = new Map<OptionsSourceType, new () => OptionsSource>();
+export abstract class OptionsSource implements IOptionsSource {
+
+	private _variableNames!: { [index: string]: string; };
+	private _type!: OptionsSourceType;
+
+	get type(): OptionsSourceType { return this._type; }
+	get variableNames(): { [index: string]: string; } { return this._variableNames; }
+
+	abstract async getOptions(): Promise<string[] | CategoryValue[]>;
+	abstract async getVariableValue(variableName: string, input: string): Promise<string>;
+	abstract getIsPassword(variableName: string): boolean;
+
+	protected constructor() {
+	}
+
+	static construct(optionsSourceType: OptionsSourceType, variableNames: { [index: string]: string }): OptionsSource {
+		const sourceConstructor = OptionsSources.get(optionsSourceType);
+		throwUnless(sourceConstructor !== undefined, loc.noOptionsSourceDefined(optionsSourceType));
+		const obj = new sourceConstructor();
+		obj._type = optionsSourceType;
+		obj._variableNames = variableNames;
+		return obj;
+	}
+}
+
+export class ArcControllersOptionsSource extends OptionsSource {
```

Comment:
So you're still making it so that the resource-deployment extension has to know about every single consumer and implement some of the logic itself? I mean at this point you might as well just have it call the azdata extension directly instead of using arc....

If we're going to do this feature it should really be more generic and something that the extensions contribute - where they can register option providers with the resource-deployment extension that are used to populate the fields. An option provider would be super generic and all the logic would be done on the contributing extension side.